website/home/_posts/wheres_the_graph.md

```diff
@@ -0,0 +1,148 @@
+---
+
+# "But Where's the Graph?!" — Why Trees Beat Graphs for LLM Workflows
+
+Look, I get it. When you think "workflow orchestration," your mind immediately goes to boxes and arrows. Some fancy graph visualization that makes your project look like a NASA mission control diagram. It's what everyone does. It's the "serious" approach.
+
+But I'm going to let you in on a little secret that transformed how we built GenSX: **trees are graphs too**, and they're dramatically easier to reason about.
+
+## The Graph Trap
+
+I used to be that developer drawing elaborate workflow diagrams on whiteboards. Building complex orchestration pipelines with explicit nodes and edges. You know, the kind that looks like this:
+
+```tsx
+const graph = new Graph()
+  .addNode("fetchHNPosts", fetchHNPosts)
+  .addNode("analyzeHNPosts", analyzePosts)
+  .addNode("generateReport", generateReport)
+  .addNode("editReport", editReport)
+  .addNode("writeTweet", writeTweet);
+
+graph
+  .addEdge(START, "fetchHNPosts")
+  .addEdge("fetchHNPosts", "analyzeHNPosts")
+  .addEdge("analyzeHNPosts", "generateReport")
+  .addEdge("generateReport", "editReport")
+  .addEdge("editReport", "writeTweet")
+  .addEdge("writeTweet", END);
+```
+
+The problem? These graph structures force you to juggle multiple mental models simultaneously. You've got the node definitions in one place, edge connections somewhere else, a separate execution flow to consider, and usually a global state object being passed around. And that's before we even get to complex patterns like branching, conditional execution, or—heaven forbid—loops.
+
+## Trees: The Elegant Simplification
+
+When we built GenSX, we made a radical bet: what if we used JSX to represent workflows as trees instead? Not because we're React fanboys (though some of us might be), but because tree structures match how humans naturally think about processes.
+
+Here's that same workflow in GenSX:
+
+```tsx
+<FetchHNPosts limit={postCount}>
+  {(stories) => (
+    <AnalyzeHNPosts stories={stories}>
+      {({ analyses }) => (
+        <GenerateReport analyses={analyses}>
+          {(report) => (
+            <EditReport content={report}>
+              {(editedReport) => (
+                <WriteTweet
+                  context={editedReport}
+                  prompt="Summarize the HN trends in a tweet"
+                />
+              )}
+            </EditReport>
+          )}
+        </GenerateReport>
+      )}
+    </AnalyzeHNPosts>
+  )}
+</FetchHNPosts>
+```
+
+"But wait," I can hear you protesting, "where's the graph?!"
+
+It's right there. A tree is just a particular kind of graph—one where each node has exactly one parent. And that constraint buys us immense clarity without sacrificing power.
+
+## When Trees Aren't Enough, Use JavaScript
+
+Trees work beautifully for most linear workflows. But what about loops, conditionals, and cycles? Those places where traditional graphs flex their muscles?
```

Comment:
Another points that might be worth sprinkling in throughout: 

You can trace a line through the execution of a graph in order to unroll it into a tree that shows the flow of execution, and how data branches and joins back together. Even non-deterministic agent workflows have starts and ends, no matter how long they run.